refactor: Header sync optimisations & simplifications

danielabrozzoni opened the pull request:
This is a partial* revival of #25968

It contains a list of most-unrelated simplifications and optimizations to the code merged in #25717:

- Avoid an IsAncestorOfBestHeaderOrTip call: Just don't call this function when it won't have any effect.
- Compute work from headers without CBlockIndex: Avoid the need to construct a CBlockIndex object just to compute work for a header, when its nBits value suffices for that. Also use some Spans where possible.
- Remove useless CBlock::GetBlockHeader: There is no need for a function to convert a CBlock to a CBlockHeader, as it's a child class of it.

It also contains the following code cleanups, which were suggested by reviewers in #25968:
- Remove redundant parameter from CheckHeadersPoW: No need to pass consensusParams, as CheckHeadersPow already has access to m_chainparams.GetConsensus()
- Remove unused parameter in ReportHeadersPresync
- Use initializer list in CompressedHeader, also make GetFullHeader const
- Use reference for chain_start in HeadersSyncState: chain_start can never be null, so it's better to pass it as a reference rather than a raw pointer

*I decided to leave out three commits that were in #25968 (4e7ac7b94d04e056e9994ed1c8273c52b7b23931, ab52fb4e95aa2732d1a1391331ea01362e035984, 7f1cf440ca1a9c86085716745ca64d3ac26957c0), since they're a bit more involved, and I'm a new contributor. If this PR gets merged, I'll comment under #25968 to note that these three commits are still up for grabs :)

sipa committed:
p2p: Avoid an IsAncestorOfBestHeaderOrTip call

Just don't call this function when it won't have any effect.

Note that we can't remove the LookupBlockIndex call, since `last_received_header`
is needed to check if new headers were received (`received_new_header`).

sipa committed:
refactor: Remove useless CBlock::GetBlockHeader

There is no need for a function to convert a CBlock to a CBlockHeader, as it's a child
class of it.

danielabrozzoni committed:
refactor: Remove redundant parameter from CheckHeadersPoW

No need to pass consensusParams, as CheckHeadersPoW already has access
to m_chainparams.GetConsensus()

Co-Authored-By: maflcko <6399679+maflcko@users.noreply.github.com>

danielabrozzoni committed:
refactor: Remove unused parameter in ReportHeadersPresync

Co-Authored-By: Aurèle Oulès <aurele@oules.com>

danielabrozzoni committed:
refactor: Use initializer list in CompressedHeader

Also mark CompressedHeader as explicit, and GetFullHeader as const

Co-Authored-By: Aurèle Oulès <aurele@oules.com>

danielabrozzoni committed:
refactor: Use reference for chain_start in HeadersSyncState

chain_start can never be null, so it's better to pass it as a reference
rather than a raw pointer

Also slightly reformat HeaderSyncState constructor to make clang-format
happy

Lastly, remove `const` from `chain_start` declaration in
headers_sync_chainwork_tests, to work aroud a false-positive
dangling-reference warning in gcc 13.0

Co-Authored-By: maflcko <6399679+maflcko@users.noreply.github.com>

ajtowns commented:
Maybe change the title to describe what's actually going on for those of us who don't memorise every pr number? :) ("Header sync optimisations/simplifications" ?)

hodlinator reviewed (commented):
Reviewed 13e123e7cc185760bb9dafc1a6892c92112e5450

Thanks for picking this up!

Conflicts somewhat with #32579 where I was aiming to change the headers-parameter to a span. But if we follow through and also take my `enum` suggestion I would still see it as an improvement on balance (see inline comment).

danielabrozzoni commented:
@hodlinator thank you for your review!

I like the idea of converting `ProcessingResult` to an enum, I think it makes the code easier to read.

As for `pow_validated_headers`, I don’t have a strong preference between using a span or an argument for I/O just yet. I’ll take a closer look at #32579 over the next few days and should have a more informed opinion afterward :) I’d also be interested to hear what other reviewers think!

Latest push:
- style improvements based on review comments

l0rinc commented on src/headerssync.h:
```diff
@@ -136,37 +134,34 @@ class HeadersSyncState {
     /** Process a batch of headers, once a sync via this mechanism has started
      *
-     * received_headers: headers that were received over the network for processing.
+     * headers: headers that were received over the network for processing.
```
nit: please reformat the block

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -85,17 +85,19 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
 
     hss.reset(new HeadersSyncState(0, Params().GetConsensus(), chain_start, chain_work));
-    (void)hss->ProcessNextHeaders({first_chain.front()}, true);
+    std::vector<CBlockHeader> to_proc{first_chain.front()};
```
There's a significant overlap with https://github.com/bitcoin/bitcoin/pull/32579/files#diff-46faf106e779941e78de61d061d7817f9dd7b1f12a4157e6991417c7f0a7e9d3L95 - would be great if we could avoid a complete rewrite after one of them is merged

danielabrozzoni replied:
Yes absolutely! I think it all comes down to https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2149379657:
- This PR changes MakeProcessNextHeaders/PopHeadersReadyForAcceptance to use headers as a I/O parameter (09c7d0f703067d455aae4ee458ba7953c29d72fb)
- #32579 changes MakeProcessNextHeaders to use span instead of vector for headers (4c45b5d349c59ed6a795819cdfbaec90bc189a24)

The reasons for switching to headers as I/O parameter are explained in sipa's commit (09c7d0f703067d455aae4ee458ba7953c29d72fb):
- It allows using the same vector storage for input and output
- Better represents that `headers` is the same vector that needs to be processed
Also, as hodlinator said in the review comment:
- It conforms to the style of net_processing.cpp (https://github.com/bitcoin/bitcoin/pull/32740/commits/09c7d0f703067d455aae4ee458ba7953c29d72fb#r2149379657)

That said, there are downsides:
- It pushes readers towards having to lean on comments more (https://github.com/bitcoin/bitcoin/pull/32740/commits/09c7d0f703067d455aae4ee458ba7953c29d72fb#r2149379657)
- It makes the tests more cumbersome (https://github.com/bitcoin/bitcoin/pull/32740/commits/09c7d0f703067d455aae4ee458ba7953c29d72fb#r2149379657)

I’m really not sure which solution is better - my C++ is still at a beginner level, so I’d really appreciate any input on this! I’m also happy to drop the first commit if reviewers feel the span approach is the better direction.

l0rinc reviewed (commented):
I'd like to review this - can you tell me how to measure the effect of the "optimizations" using either the microbenchmarks or a reindex or anything that shows the magnitude of change?
Would be helpful if you could publish your measurements if you have any.

l0rinc commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -85,7 +85,7 @@ BOOST_AUTO_TEST_CASE(headers_sync_state)
             ArithToUint256(1), Params().GenesisBlock().nBits);
 
-    const CBlockIndex* chain_start = WITH_LOCK(::cs_main, return m_node.chainman->m_blockman.LookupBlockIndex(Params().GenesisBlock().GetHash()));
+    const CBlockIndex& chain_start = {*Assert(WITH_LOCK(::cs_main, return m_node.chainman->m_blockman.LookupBlockIndex(Params().GenesisBlock().GetHash())))};
```
It's a bit weird to have `Assert` outside, how about:
```suggestion
    const CBlockIndex& chain_start = WITH_LOCK(::cs_main, return *Assert(m_node.chainman->m_blockman.LookupBlockIndex(Params().GenesisBlock().GetHash())));
```

ajtowns replied:
Returning a pointer from an anonymous function with no explicit return type seems safer and easier to follow than returning a reference? Not sure what's weird about having the Assert outside the lock. We have a similar instance already:

https://github.com/bitcoin/bitcoin/blob/eb137184482cea8a2a7cdfcd3c73ec6ae5fdd0d6/src/node/blockstorage.cpp#L935

l0rinc replied:
I prefer validating closely to the source, but I don't mind either way, thanks for the comment

danielabrozzoni replied:
Thanks for catching it and sorry for not specifying in my push comment!

I dropped it on purpose because it caused a compilation warning:

```
[ 73%] Building CXX object src/test/CMakeFiles/test_bitcoin.dir/headers_sync_chainwork_tests.cpp.o
/home/daniela/Developer/BitcoinCore/bitcoin/src/test/headers_sync_chainwork_tests.cpp:55:37: warning: predefined identifier is only valid inside function [-Wpredefined-identifier-outside-function]
   55 |     const CBlockIndex& chain_start{*Assert(WITH_LOCK(::cs_main, return m_node.chainman->m_blockman.LookupBlockIndex(genesis.GetHash())))};
      |                                     ^
/home/daniela/Developer/BitcoinCore/bitcoin/src/util/check.h:115:75: note: expanded from macro 'Assert'
  115 | #define Assert(val) inline_assertion_check<true>(val, __FILE__, __LINE__, __func__, #val)
      |                                                                           ^
```
This didn't happen before my rebase onto #32579, as chain_start was initialized inside a function.

I also had a preference for using the Assert, but I'm not sure if it can still be done somehow... I'm open to suggestions :)

l0rinc replied:
Can you try my originally [suggestion](https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2226981997) to put the `Assert` inside?
```C++
const CBlockIndex& chain_start{WITH_LOCK(::cs_main, return *Assert(m_node.chainman->m_blockman.LookupBlockIndex(genesis.GetHash())))};
```
it works for me locally. Or ignore, since we've already acked.

l0rinc reviewed (changes requested):
Concept ACK, I'm not sure about the first change but the rest make sense.

reviewed f4b33a14f27bb3ee734e8d4eb9c2b2ccaa5df52a, left some comments for most commits, my main concern is the commit where we're reusing the headers vector, making the code significantly more contrived.
It would also be great if we could extract the common parts of this and @hodlinator's change and merge that (or his change) before we do this.

danielabrozzoni commented:
After some thinking, I decided to drop the first commit, and let ProcessNextHeaders take the headers argument and return it, instead of using it as a I/O argument. I tried to write a benchmark to measure how much using the vector as I/O argument would optimize the code, but it convinced me that the new interface would be too hard to work with, as reviewers pointed out. While it might still be a performance improvement, I'm not sure if it's really worth it.

For anyone interested, here's my attempt at writing a benchmark: 852aca5e34320449c47488d181e98ce72b74adfd and [my branch](https://github.com/danielabrozzoni/bitcoin/commits/upforgrabs/25968-bench/)
(Unfortunately, since ProcessNextHeaders changes the content of the headers vector passed in, I had to copy the vector every single time, which I suppose could make the benchmark a bit unreliable? There are also a few assertions in the bench.run() function, which I used to make sure that my code wasn't bugged, but I'm not sure if they should be there in any serious attempt at benchmarking)

Will update the PR in the next few days :)

danielabrozzoni committed:
refactor: Use std::span in HasValidProofOfWork

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

danielabrozzoni commented on src/validation.h:
```diff
@@ -394,7 +394,7 @@ bool TestBlockValidity(BlockValidationState& state,
 
 /** Check with the proof of work on each blockheader matches the value in nBits */
-bool HasValidProofOfWork(const std::vector<CBlockHeader>& headers, const Consensus::Params& consensusParams);
+bool HasValidProofOfWork(const std::span<const CBlockHeader> headers, const Consensus::Params& consensusParams);
```
That makes a lot of sense, thank you!
I only removed the const, I prefer to keep the argument names

danielabrozzoni commented on src/validation.cpp:
```diff
@@ -4159,9 +4159,9 @@ std::vector<unsigned char> ChainstateManager::GenerateCoinbaseCommitment(CBlock&
 }
 
-bool HasValidProofOfWork(const std::vector<CBlockHeader>& headers, const Consensus::Params& consensusParams)
+bool HasValidProofOfWork(const std::span<const CBlockHeader> headers, const Consensus::Params& consensusParams)
```
Thanks, I removed the const and put this change in a separate commit

danielabrozzoni reviewed (commented):
In my last push (f33bd2f):
- Removed first commit as per my last comment (https://github.com/bitcoin/bitcoin/pull/32740#issuecomment-3144999375)
- Addressed review comments

l0rinc commented on src/validation.cpp:
```diff
@@ -4149,8 +4149,8 @@ std::vector<unsigned char> ChainstateManager::GenerateCoinbaseCommitment(CBlock&
 {
-    return std::all_of(headers.cbegin(), headers.cend(),
-            [&](const auto& header) { return CheckProofOfWork(header.GetHash(), header.nBits, consensusParams);});
+    return std::ranges::all_of(headers,
```
nit: this change could either be merged with the [vector-to-std::span](https://github.com/bitcoin/bitcoin/pull/32740/commits/dac0af8fb03a7ba39a36b5f7fa204683ae146013) or moved over to that commit, it's not needed in 80c63c9ceccc89295f004099c8cd7e7dfd99b556

l0rinc commented on src/primitives/block.h:
```diff
@@ -101,18 +101,6 @@ class CBlock : public CBlockHeader
         m_checked_merkle_root = false;
     }
 
-    CBlockHeader GetBlockHeader() const
```
nice!

l0rinc commented on src/headerssync.cpp:
```diff
@@ -24,14 +24,13 @@ constexpr size_t REDOWNLOAD_BUFFER_SIZE{14827}; // 14827/624 = ~23.8 commitments
-    m_current_chain_work(chain_start->nChainWork),
-    m_last_header_received(m_chain_start->GetBlockHeader()),
-    m_current_height(chain_start->nHeight)
+                                   const CBlockIndex& chain_start, const arith_uint256& minimum_required_work) : m_commit_offset(FastRandomContext().randrange<unsigned>(HEADER_COMMITMENT_PERIOD)),
```
nit: if you touch again, please format this differently

danielabrozzoni replied:
I would still keep the reformatting as `clang-format` doesn't like the one on master, and reformats it as I previously did (with the member initializers pushed to the right). At least this reformat is nice and `clang-format` doesn't try to modify it. 

I hadn't considered `git blame` though, I don't love the idea of polluting it...

hodlinator replied:
Given the current stats on initializer list formatting, I don't think there's enough reason to open a PR to turn the tide by changing src/.clang-format.

```shell
# BeforeColon (current default in src/.clang-format, style partially objected to in this thread):
$ grep -zoP "\n( {4})+( {2})m_\w+[({]" -r src/ |wc -l
135

# AfterColon (what I expected to be most common):
$ grep -zoP "\n( {4})+m_\w+[({]"  -r src/ |wc -l
117

# BeforeComma (BetaMax of initializer lists, leads to smallest diffs):
$ grep -zoP "\n( {4})+, m_\w+[({]"  -r src/ |wc -l
17
```

https://releases.llvm.org/16.0.0/tools/clang/docs/ClangFormatStyleOptions.html#breakconstructorinitializers

So this thread can be resolved IMO. Cheers to @l0rinc for the reminder.

l0rinc reviewed (approved):
code review ACK f33bd2f1425050ff89caa018edfb2265a754fb5d

nit: the PR isn't really an optimization anymore, consider updating the title

hodlinator reviewed (commented):
Reviewed f33bd2f1425050ff89caa018edfb2265a754fb5d

hodlinator commented on src/net_processing.cpp:
```diff
@@ -2937,7 +2937,7 @@ void PeerManagerImpl::ProcessHeadersMessage(CNode& pfrom, Peer& peer,
-        if (IsAncestorOfBestHeaderOrTip(last_received_header)) {
+        if (!already_validated_work && IsAncestorOfBestHeaderOrTip(last_received_header)) {
             already_validated_work = true;
         }
```
(re https://github.com/bitcoin/bitcoin/pull/32740#pullrequestreview-3102319617:
> nit: the PR isn't really an optimization anymore, consider updating the title

I guess this is technically still an optimization).

danielabrozzoni committed:
refactor: Compute work from headers without CBlockIndex

Avoid the need to construct a CBlockIndex object just to compute work for a header,
when its nBits value suffices for that.

Co-Authored-By: Pieter Wuille <pieter@wuille.net>

danielabrozzoni reviewed (commented):
Thanks for the reviews!

In my last push:
- Rebased onto the right commit :sweat_smile: (see https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2269846966)
- Addressed review comments

hodlinator reviewed (approved):
ACK 50e61f448eedfe0ccfa07f3124aeef9c7762a69b

PR improves the code in several minor ways, such as:
* Computing PoW for a `CBlockHeader` (or `CBlock`) without having to jump through temporarily constructed `CBlockIndex` (552b9c3a565c1817074468bf71fb4526f3a47f42).
* Removing `CBlock::GetBlockHeader()` leaves less room for forgetting to properly set fields in the new `CBlockHeader` instance (67aa62c80e7d2faa485ba020cefe262f479190f4). 
* Changes `const CBlockIndex* HeadersSyncState::m_chain_start` to a reference as it is never null and never changed (50e61f448eedfe0ccfa07f3124aeef9c7762a69b).

l0rinc commented:
recrACK 50e61f448eedfe0ccfa07f3124aeef9c7762a69b

I only left nits, I'm fine with merging it as is.

danielabrozzoni commented on src/headerssync.cpp:
```diff
@@ -41,7 +44,7 @@ HeadersSyncState::HeadersSyncState(NodeId id, const Consensus::Params& consensus
     // chain to be longer than this (at the current time -- in the future we
     // could try again, if necessary, to sync a longer chain).
-    m_max_commitments = 6*(Ticks<std::chrono::seconds>(NodeClock::now() - NodeSeconds{std::chrono::seconds{chain_start->GetMedianTimePast()}}) + MAX_FUTURE_BLOCK_TIME) / HEADER_COMMITMENT_PERIOD;
+    m_max_commitments = 6 * (Ticks<std::chrono::seconds>(NodeClock::now() - NodeSeconds{std::chrono::seconds{chain_start.GetMedianTimePast()}}) + MAX_FUTURE_BLOCK_TIME) / HEADER_COMMITMENT_PERIOD;
```
I'm happy to put this in draft and wait for #32579 to be merged first... This PR is smaller, and probably easier to rebase. What do you think?

l0rinc replied:
I personally would let the maintainers decide the order, ank keep both open

polespinasa commented on src/headerssync.h:
```diff
@@ -31,16 +31,17 @@ struct CompressedHeader {
     }
 
-    CompressedHeader(const CBlockHeader& header)
+    explicit CompressedHeader(const CBlockHeader& header)
```
I don't get this change, what's the use-case for explicit here? and why is better to use the list instead of a "normal" constructor?

danielabrozzoni replied:
The explicit was suggested in https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2226976728. As far as I know it's good practice, so that the compiler can't make implicit conversions to resolve the parameters of a function. Also see: https://github.com/bitcoin/bitcoin/blob/292ea0eb8982faef460c210bd4215d603f487463/doc/developer-notes.md?plain=1#L835

The initializer list was suggested in https://github.com/bitcoin/bitcoin/pull/25968#pullrequestreview-1096400728, I think it's consistent with the rest of the code, but I'm not sure if there's any benefit to it

danielabrozzoni replied:
The explicit was suggested in https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2226976728. As far as I know it's good practice, so that the compiler can't make implicit conversions to resolve the parameters of a function. Also see: https://github.com/bitcoin/bitcoin/blob/292ea0eb8982faef460c210bd4215d603f487463/doc/developer-notes.md?plain=1#L835

The initializer list was suggested in https://github.com/bitcoin/bitcoin/pull/25968#pullrequestreview-1096400728, I think it's consistent with the rest of the code, but I'm not sure if there's any benefit to it


Thanks for the review btw! I'm currently working on rebasing, it might take a couple of days though (I have a test failing...)

polespinasa reviewed (approved):
Code Review  ACK 50e61f448eedfe0ccfa07f3124aeef9c7762a69b
(will re-review after rebasing)

This PR cleans and simplifies the code in many ways like avoiding calling unneeded functions, avoiding constructing temporary objects, removing unused params, among others...

Just left a single comment with a question :)

btw 552b9c3a565c1817074468bf71fb4526f3a47f42 and 67aa62c80e7d2faa485ba020cefe262f479190f4 are pretty :)

hodlinator reviewed (approved):
re-ACK 8a4ab45c13dd7ca4474bc4621110da82e0a634f8

Changes since previous review (https://github.com/bitcoin/bitcoin/pull/32740#pullrequestreview-3128785887):

* More helpful commit messages.
* Rebased, especially to handle #32579.

danielabrozzoni commented:
@hodlinator thanks! Was about to comment with the update haha :)

In my last push:

- Rebase after #32579 
- Address https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2289739420, https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2289729638, https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2289725133
- Specify in 8a4ab45c13dd7ca4474bc4621110da82e0a634f8 commit message that I'm also slightly reformatting the code

hodlinator reviewed (approved):
re-ACK 9a387e9175d34d1eaf83b9008f5e857d1eb04b8e

Changes since previous review (https://github.com/bitcoin/bitcoin/pull/32740#pullrequestreview-3409949311):

* Re-added `Assert` in response to https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2485972867
  I was slightly worried the lock lambda would become more likely to return by-value, instead of by-reference, but `CBlockIndex` has `protected`/`delete`d copying:
  https://github.com/bitcoin/bitcoin/blob/9a387e9175d34d1eaf83b9008f5e857d1eb04b8e/src/chain.h#L335-L349

danielabrozzoni commented:
Hey, sorry this took a while. I tried rebasing onto #33785 but it wouldn't fix the warning. 

In my last push (de4242f):
- I didn't rebase, to hopefully make re-reviewing easier
- I removed `const` as suggested in https://github.com/bitcoin/bitcoin/pull/32740#discussion_r2509736434, I like it a bit better than the previous solution
- I squashed the fix inside the last commit instead pushing as a separate one

> Technically I think the repo policy is that every commit should build on all CI,

Understood, thanks for explaining! :)

hodlinator commented on src/test/headers_sync_chainwork_tests.cpp:
```diff
@@ -52,7 +52,7 @@ constexpr size_t COMMITMENT_PERIOD{600}; // Somewhat close to mainnet.
 struct HeadersGeneratorSetup : public RegTestingSetup {
     const CBlock& genesis{Params().GenesisBlock()};
-    const CBlockIndex& chain_start{WITH_LOCK(::cs_main, return *Assert(m_node.chainman->m_blockman.LookupBlockIndex(genesis.GetHash())))};
+    const CBlockIndex& chain_start{ChainStart()};
```
Aha, that explains why I was getting other results when not rebasing onto latest master (https://github.com/bitcoin/bitcoin/pull/32740#issuecomment-3559658957).

hodlinator reviewed (approved):
re-ACK de4242f47476769d0a7f3e79e8297ed2dd60d9a4

Only removed `const` from `HeadersGeneratorSetup::chain_start` to work around GCC 13 issue since my previous ACK (https://github.com/bitcoin/bitcoin/pull/32740#pullrequestreview-3410905600).

achow101 commented:
ACK de4242f47476769d0a7f3e79e8297ed2dd60d9a4
